### 1. Utilize o método da bissecção para encontrar uma solução com precisão de $10^{−5}$ para os seguintes problemas:

##### (b) $x \cos(x) − 2x^2 + 3x − 1 = 0$ para $0.2 ≤ x ≤ 0.3$ e $1.2 ≤ x ≤ 1.3$.

In [1]:
from math import cos

def funcao(x):
    return x * cos(x) - 2*x**2 + 3*x - 1

# a e b são os intervalos, f é a função para a qual queremos encontrar a raiz
def bissecao(a, b, f, precisao, max_iteracoes):
    for k in range(0, max_iteracoes):
        p = (a+b)/2 # a possível raiz é o ponto médio entre a e b
        fp = f(p) # pega o valor da possível raiz
        if fp == 0 or abs(fp) < precisao: # se f(p) estiver dentro da precisão, p é um 0
            return p
        if f(a) * fp < 0: # se houver 0 no intervalo [a,p] então [a,p] é o novo intervalo
            b = p
        else: # caso contrário, [p, b] é o novo intervalo
            a = p
    return f'não foi encontrada uma raiz de f no intervalo com precisão de {precisao} e \
            número de iterações {max_iteracoes}'

print(f'Raiz de f(x) no intervalo [0.2, 0.3]: {bissecao(0.2, 0.3, funcao, 10**-5, 100)}')
print(f'Raiz de f(x) no intervalo [1.2, 1.3]: {bissecao(1.2, 1.3, funcao, 10**-5, 100)}')

Raiz de f(x) no intervalo [0.2, 0.3]: 0.29752807617187504
Raiz de f(x) no intervalo [1.2, 1.3]: 1.256622314453125


### 3. Para cada uma das equações a seguir determine uma função de iteração φ(x) e um intervalo [a, b] no qual as iterações convergirão para uma solução positiva da equação:
##### (a) $3x^2 − e^x = 0$
##### Encontre as soluções com precisão de $10^{−2}$ usando o método do ponto fixo com a função de iteração escolhida.

Definindo a função no Python

In [2]:
from math import exp

def funcao_3(x):
    return 3*x**2 - exp(x)

Escolhendo o intervalo

In [3]:
print(funcao_3(3))
print(funcao_3(4))

6.914463076812332
-6.598150033144236


O intervalo [3, 4] tem uma raiz de $f(x)$ (sinal muda).

Agora é preciso definir a função de iteração.
$$
3x^2 - e^x = 0 \\
e^x = 3x^2 \\
g(x) = \ln(3x^2) \\
$$
A derivada da função é
$$
g'(x) = \frac{2}{x}
$$
E ela tem que ser menor do que 1 em módulo em todo o intervalo $3 < x < 4$, o que claramente é verdade.


Definindo a função $g(x)$ em Python

In [4]:
from math import log

def g(x):
    return log(3*x**2)

Aplicando o método do ponto fixo

In [5]:
# x_0 é o palpite inicial, g é a função de iteração 
# e f é a função para a qual queremos encontrar a raiz
def ponto_fixo(x_0, g, f, precisao, max_iteracoes):
    x = g(x_0)
    for i in range(max_iteracoes):
        if abs(f(x)) < precisao: # se f(x) estiver perto o suficiente de 0, retorna x
            return x
        x = g(x) # utiliza a função de iteração g
    return f'não foi possível encontrar o zero da função com precisao de {precisao}'

print(ponto_fixo(3.5, g, funcao_3, 10**-2, 100))

3.7325920355902507


Então 3.732592 é aproximadamente um zero da função.

##### (b) x − cos(x) = 0

Definindo a função e o intervalo:

In [6]:
from math import cos

def funcao_3_b(x):
    return x-cos(x)

print(funcao_3_b(0))
print(funcao_3_b(1))

-1.0
0.45969769413186023


Então há um zero da função no intervalo [0, 1]. Definindo a função iterativa $g(x)$:
$$
g(x) = \cos(x)
$$

E a sua derivada, $-\mathrm{sen}(x)$ tem ser ser menor do que $1$ em módulo no intervalo, o que é verdade.

Basta agora usar a função do ponto fixo definida na letra a) com os novos parâmetros para encontrar a resposta:

In [7]:
from math import cos

def g(x):
    return cos(x)

print(ponto_fixo(0, g, funcao_3_b, 10**-2, 100))

0.7442373549005569


Então 0.744237 é aproximadamente um zero da função.